### PR TITLE
Add support for HTTP DELETE method

### PR DESCRIPTION
**Pull request by biasedbit**
Adds support for HTTP DELETE method via `@Delete` annotation.

### CODE REVIEW

File: misk/src/test/kotlin/misk/web/actions/SupportedHttpMethodsTest.kt [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+import misk.web.mediatype.MediaTypes
+import okhttp3.OkHttpClient
+import okhttp3.Request
+import okhttp3.RequestBody.Companion.toRequestBody
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.Test
+import java.util.UUID
+import javax.inject.Inject
+
+// Tests all supported HTTP methods.
```
Comment (biasedbit) [inline misk/src/test/kotlin/misk/web/actions/SupportedHttpMethodsTest.kt]: Couldn't seem to find a similar test to add the DELETE test to. No strong reason to test GET and POST (since there's a bunch of other code already testing it), just erring on the side of having a test for all HTTP methods.

Happy to drop GET/POST tests if y'all feel it's redundant.

File: misk/src/main/kotlin/misk/web/jetty/WebActionsServlet.kt [hunk 2]
```diff
@@ -82,7 +86,7 @@ internal class WebActionsServlet @Inject constructor(
       val candidateActions = boundActions.mapNotNull {
         it.match(httpCall.dispatchMechanism, requestContentType, requestAccepts, httpCall.url)
       }
-      val bestAction = candidateActions.sorted().firstOrNull()
+      val bestAction = candidateActions.min()
```
Comment (biasedbit) [inline misk/src/main/kotlin/misk/web/jetty/WebActionsServlet.kt]: I don't know what's our stance on minor, somewhat-unrelated, cosmetic changes (no contribution guidelines) so I threw this in (separate commit). Happy to drop/yank into separate PR.

File: misk/src/main/kotlin/misk/web/actions/WebActionFactory.kt [hunk 3]
```diff
@@ -40,11 +41,12 @@ internal class WebActionFactory @Inject constructor(
     webActionClass: KClass<A>,
     pathPrefix: String = "/"
   ): List<BoundAction<A>> {
-    // Find the function with Get, Post, or ConnectWebSocket annotation. Only one such function is
-    // allowed.
+    // Find the function with Get, Post, Delete or ConnectWebSocket annotation.
+    // Only one such function is allowed.
     val actionFunctions = webActionClass.functions.mapNotNull {
       if (it.findAnnotationWithOverrides<Get>() != null ||
```
Comment (biasedbit) [inline misk/src/main/kotlin/misk/web/actions/WebActionFactory.kt]: Wouldn't be a terrible idea to turn this into a loop over a list but a) I'll likely be adding support for other methods (PUT/PATCH) soon, and b) might as well make that change as its own PR.

File: misk/src/main/kotlin/misk/web/DispatchMechanism.kt [hunk 4]
```diff
@@ -8,8 +8,9 @@ import org.eclipse.jetty.http.HttpMethod
  */
 enum class DispatchMechanism {
```
Comment (biasedbit) [inline misk/src/main/kotlin/misk/web/DispatchMechanism.kt]: Grouping HTTP methods together in this enum was done as a separate commit. Trivial to drop if we feel strongly about alphabetical order.

File: misk/src/main/kotlin/misk/web/jetty/WebActionsServlet.kt (hunk 2 above)
Comment (swankjesse) [reply misk/src/main/kotlin/misk/web/jetty/WebActionsServlet.kt]: Tidying up along the way is nice.

File: misk/src/test/kotlin/misk/web/actions/SupportedHttpMethodsTest.kt (hunk 1 above)
Comment (swankjesse) [reply misk/src/test/kotlin/misk/web/actions/SupportedHttpMethodsTest.kt]: I think the most useful thing to test here is whether a body is required or expected. Could test that GET declarations can't have a body, and we should decide whether we permit/require bodies for delete.

File: misk/src/test/kotlin/misk/web/actions/SupportedHttpMethodsTest.kt [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        .build()
+
+    val response = httpClient.newCall(request).execute()
+    assertThat(response.isSuccessful).isTrue()
+    assertThat(response.body?.string()).isEqualTo(content)
+  }
+
+  @Test
+  fun delete() {
+    val content = UUID.randomUUID().toString()
```
Comment (swankjesse) [inline misk/src/test/kotlin/misk/web/actions/SupportedHttpMethodsTest.kt]: Please, nothing random in tests. It brings about flakiness.

File: misk/src/test/kotlin/misk/web/actions/SupportedHttpMethodsTest.kt [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+      install(WebActionModule.create<PostAction>())
+      install(WebActionModule.create<DeleteAction>())
+    }
+  }
+
+  internal class GetAction @Inject constructor() : WebAction {
+    @Get(PATH)
+    fun get(): String = CONTENT
+    companion object {
+      const val PATH = "/resources/id"
```
Comment (swankjesse) [inline misk/src/test/kotlin/misk/web/actions/SupportedHttpMethodsTest.kt]: My preference is to avoid the constants and just use strings here. The benefits of the indirection don't pay for themselves.

In particular we really want to be able to copy & paste test code without refactoring other tests. Duplication across tests is okay; it means each test is more self-contained.

File: misk/src/test/kotlin/misk/web/actions/SupportedHttpMethodsTest.kt (hunk 5 above)
Comment (biasedbit) [reply misk/src/test/kotlin/misk/web/actions/SupportedHttpMethodsTest.kt]: Texas-size 10-4; will change.

File: misk/src/test/kotlin/misk/web/actions/SupportedHttpMethodsTest.kt (hunk 6 above)
Comment (biasedbit) [reply misk/src/test/kotlin/misk/web/actions/SupportedHttpMethodsTest.kt]: Hard agreement and will do!

File: misk/src/test/kotlin/misk/web/actions/SupportedHttpMethodsTest.kt (hunk 1 above)
Comment (biasedbit) [reply misk/src/test/kotlin/misk/web/actions/SupportedHttpMethodsTest.kt]: Can do. The spec doesn't explicitly forbid message body for DELETE requests (okhttp also seems to accept one in the `RequestBuilder`, though it defaults to empty), but the [section on Message Body](https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.3) has something on ignoring them when there's no defined semantics:

> A server SHOULD read and forward a message-body on any request; if the request method does not include defined semantics for an entity-body, then the message-body SHOULD be ignored when handling the request.

I recall coming across libraries that don't even allow specifying a body so I don't mind being opinionated about disallowing body for DELETEs. How do you reckon it's the best way to disallow content? Validations on `WebActionServlet#doDelete`, prior to delegating to `#handleCall`?

Comment (swankjesse) [reply misk/src/test/kotlin/misk/web/actions/SupportedHttpMethodsTest.kt]: I think probably we should just forbid DELETE actions from declaring a request body. That way we blame the programmer for making a mistake, and don’t worry too much about what crazy internet users send us.

Maybe we do that in WebActionBinding.newWebActionBinder(). That can be a follow-up PR!
